src/stream.rs

```diff
@@ -15,4 +17,48 @@ impl AnyStream {
     pub fn new<S: Stream + 'static>(stream: S) -> Self {
         Self(Box::pin(stream))
     }
+
+    /// Transfers at least one byte from the stream into the read buffer.
+    pub async fn read(&mut self, read_buffer: &mut BytesMut) -> Result<(), StreamError> {
+        let byte_count = self.0.read_buf(read_buffer).await?;
+
+        if byte_count == 0 {
+            // The result is 0 if the stream reached "end of file" or the read buffer was already
+            // full before calling `read_buf`. Because we use an unlimited buffer we know that
+            // the first case occurred.
+            Err(StreamError::Closed)
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Transfers all bytes from the write buffer into the stream.
+    pub async fn write_all(&mut self, write_buffer: &mut BytesMut) -> Result<(), StreamError> {
+        while !write_buffer.is_empty() {
+            let byte_count = self.0.write_buf(write_buffer).await?;
```

Comment:
Previously we were using [`write_all_buf`](https://docs.rs/tokio/latest/tokio/io/trait.AsyncWriteExt.html#method.write_all_buf). According to its documentation there is no way to detect a closed stream. So I decided to use a `while` + [`write_buf`](https://docs.rs/tokio/latest/tokio/io/trait.AsyncWriteExt.html#method.write_buf).